Framework/Geometry/inc/MantidGeometry/Objects/Track.h

```diff
@@ -186,6 +186,7 @@ class MANTID_GEOMETRY_DLL Track {
   Track();
   /// Constructor
   Track(const Kernel::V3D &startPt, const Kernel::V3D &unitVector);
+  virtual ~Track() = default;
```

Comment:
Could I check on the requirement that this becomes virtual? As the class had no virtual functions before it could omit the vptr and keep the objects smaller and remove indirection when calling the methods?

Reply:
Danny can probably explain it better. But from what I saw a small Mock version of the class has been constructed for testing, with a single overridden virtual method. This mock version, call it MockTrack, is passed around in a pointer to Track, therefore motivating the creation of a virtual destructor. In both cases the destructors are default, so it’s possible you could omit the virtual declaration I guess?

Reply:
Ah yes sorry I'd missed the new mock type. You're right that this is mainly a performance concern but of course not rooted in data and it is difficult to quantify. Do you know if there is any measureable performance change in now allocating the tracks on the heap as opposed to the stack? Typically memory allocation/virtual function lookup are the slow things but it is probably worth mentioning
This is the perennial problem I suppose. For the choice to be made at runtime then the they will have to be virtual. For non-virtual methods then mocking requires some other [injection method](https://chromium.googlesource.com/external/github.com/google/googletest/+/HEAD/googlemock/docs/cook_book.md#mockingnonvirtualmethods) either a template or some factory function typically.
As you say maybe that is too much effort for what it is wortth for the unit test?

Reply:
I've just done some performance tests using the following script with the Track class having a virtual\non-virtual method and destructor:
```
# import mantid algorithms, numpy and matplotlib
from mantid.simpleapi import *
import matplotlib.pyplot as plt
import numpy as np

data = CreateSampleWorkspace(WorkspaceType='Histogram', NumBanks=1, BankPixelWidth=5, XMin=1000)
data = ConvertUnits(data, Target='Wavelength')
SetSample(data, Geometry={'Shape': 'Cylinder', 'Height': 5.0, 'Radius': 1.0,
                  'Center': [0.0,0.0,0.0]},
                Material={'ChemicalFormula': 'V'})

abscor = MonteCarloAbsorption(data, EventsPerPoint=10000,
                              ResimulateTracksForDifferentWavelengths=True)
# simulated wavelengths are 0.29, 1.52, 2.76, 4.00 and 5.24 angstroms
corrected = data/abscor
```
The script takes just under 20s to run in both cases on my laptop on a release build so no noticeable difference. I think the script is creating ~47.5 million tracks (10,000 scenarios x 25 histograms x 95 bins x 2 for before\after). 

I've spent some time trying to optimize the track\surface intercept code as part of my work on multiple scattering and the limiting factor is the time spent doing Geometry::IObject::interceptSurface. So I guess either any change to the virtual status of the Track is insignificant compared to that or else the virtual behaviour is removed by the optimizer since there is no child class below Track in the Geometry target (only in the AlgorithmsTest target).

On that basis are you happy to go with the changes as they are?

Reply:
Thanks for doing the performance checks - it's good to get some proper data to confirm this. I agree there is no point in complicating the code if the difference is not measurable. It certainly sounds like `interceptSurface` is masking any change from it. I'm happy with the changes now we checked that.